libtransmission/global-ip-cache.cc

```diff
@@ -0,0 +1,363 @@
+// This file Copyright © 2023-2023 Mnemosyne LLC.
+// It may be used under GPLv2 (SPDX: GPL-2.0-only), GPLv3 (SPDX: GPL-3.0-only),
+// or any future license endorsed by Mnemosyne LLC.
+// License text can be found in the licenses/ folder.
+
+#include <algorithm> // std::all_of
+#include <chrono>
+#include <cstddef>
+#include <string_view>
+#include <utility> // std::move
+
+#ifdef _WIN32
+#include <ws2tcpip.h>
+#else
+#include <cerrno>
+#include <sys/socket.h>
+#endif
+
+#include <fmt/core.h>
+
+#include "libtransmission/log.h"
+#include "libtransmission/global-ip-cache.h"
+#include "libtransmission/tr-assert.h"
+#include "libtransmission/utils.h"
+
+namespace
+{
+
+using namespace std::literals;
+
+auto constexpr IPQueryServices = std::array{ "https://icanhazip.com"sv, "https://api64.ipify.org"sv };
```

Comment:
Didn't we set up a couple of own hosts for this?..

Reply:
It will be a few minutes work for me to change the code to use your IP query endpoints if you do have them.

Reply:
We have these endpoints set up
* https://ip4.transmissionbt.com/
* https://ip6.transmissionbt.com/

Though we could also expose a dual-stack hostname if that is prefered

Reply:
@titer A dual-stack hostname would be most convenient.

Does that take a lot more work for you?

Reply:
On a second thought, the current endpoints are fine. It does not take a lot of code change on my part to use them.

Please keep them as they are.